File: src/01_preparar_dataset.py
"""
01_preparar_dataset.py
----------------------
Lee Calculadora 2024.sav, filtra CIRUGIA=1, selecciona las 13 variables del
modelo publicado + outcome (TODOSCA), construye dummies de TSH y guarda el
dataset analítico como CSV (para Python) y RDS-compatible vía feather (opcional).

Ejecutar desde la carpeta del proyecto:
    python analisis/01_preparar_dataset.py
"""

from pathlib import Path
import pandas as pd
import pyreadstat

ROOT = Path(__file__).resolve().parent.parent
SAV  = ROOT / "Calculadora 2024.sav"
OUT  = ROOT / "analisis" / "datos_modelo.csv"

# ---------------------------------------------------------------------------
# 1. Cargar
# ---------------------------------------------------------------------------
df, meta = pyreadstat.read_sav(str(SAV))
print(f"Cargado: {df.shape[0]} pacientes × {df.shape[1]} variables")

# ---------------------------------------------------------------------------
# 2. Filtrar cohorte: solo pacientes con cirugía (gold standard = AP definitiva)
# ---------------------------------------------------------------------------
df = df[df["CIRUGIA"] == 1].copy()
print(f"Tras filtrar CIRUGIA=1: {df.shape[0]} pacientes")

# ---------------------------------------------------------------------------
# 3. Variables del modelo publicado (Tabla 1)
#    Outcome: TODOSCA  (0 benigno / 1 cáncer)
# ---------------------------------------------------------------------------
out  = "TODOSCA"
predictores = [
    "AFCDT",              # Family history of TC
    "SEX",                # Gender (1 = male)
    "EDAD",               # Age
    "EDADCUADRADO",       # Squared age
    "TSHSUPRIMIDA",       # TSH 0–0.369
    "TSHALTA",            # TSH > 4.701
    "TIROIDITIS",         # Autoimmune thyroiditis
    "SOLIDO1",            # Solid nodule
    "GANGLIOSOSPECHOSO1", # Suspicious adenopathies
    "HIPOECOICO1",        # Hypoechoic
    "IRREGULAR1",         # Margins microlobed/irregular
    "MACROCA1",           # Macrocalcifications
    "MICROCA1",           # Microcalcifications
    "FORMA1",             # Taller than wide (1 = más alto que ancho)
]

# Renombrar a etiquetas en inglés (más cómodo para el código y la publicación)
rename = {
    "AFCDT":              "fhx_tc",
    "SEX":                "male",
    "EDAD":               "age",
    "EDADCUADRADO":       "age2",
    "TSHSUPRIMIDA":       "tsh_low",
    "TSHALTA":            "tsh_high",
    "TIROIDITIS":         "autoimmune",
    "SOLIDO1":            "solid",
    "GANGLIOSOSPECHOSO1": "susp_nodes",
    "HIPOECOICO1":        "hypoechoic",
    "IRREGULAR1":         "irregular",
    "MACROCA1":           "macrocalc",
    "MICROCA1":           "microcalc",
    "FORMA1":             "taller_than_wide",
    "TODOSCA":            "cancer",
    "AÑOCIRUGIA":         "year_surgery",
}

dat = df[predictores + [out, "AÑOCIRUGIA"]].rename(columns=rename)

# Corregir typo evidente: año 2921 → 2021
dat.loc[dat["year_surgery"] == 2921, "year_surgery"] = 2021

# Recalcular age2 desde age (en SPSS quedaron 14 NA aunque age estaba completo)
dat["age2"] = dat["age"] ** 2

# ---------------------------------------------------------------------------
# 4. Comprobar tipos y missings
# ---------------------------------------------------------------------------
print("\nMissings por variable:")
print(dat.isna().sum().sort_values(ascending=False))

# Casos completos (el artículo original presumiblemente usó complete-case)
n_pre = len(dat)
dat = dat.dropna(subset=["cancer"])  # outcome obligatorio
print(f"\nTras eliminar outcome NA: {len(dat)} (había {n_pre})")

# Tipos: outcome y categóricas como int para sklearn/ranger
for c in dat.columns:
    if c not in ("age", "age2", "year_surgery"):
        dat[c] = dat[c].astype("Int64")  # mantiene NA
dat["year_surgery"] = dat["year_surgery"].astype("Int64")

# ---------------------------------------------------------------------------
# 5. Resumen rápido
# ---------------------------------------------------------------------------
print("\nDistribución del outcome:")
print(dat["cancer"].value_counts(dropna=False))
print(f"Prevalencia de cáncer: {dat['cancer'].mean()*100:.1f}%")

print("\nMissings tras limpieza:")
miss = dat.isna().sum()
print(miss[miss > 0] if miss.sum() > 0 else "  (ninguno)")

# ---------------------------------------------------------------------------
# 6. Guardar dos versiones: completa (con NAs) y complete-case
# ---------------------------------------------------------------------------
dat.to_csv(OUT, index=False)
print(f"\nGuardado completo:      {OUT}  ({len(dat)} filas × {len(dat.columns)} columnas)")

OUT_CC = ROOT / "analisis" / "datos_modelo_cc.csv"
# Complete-case sobre las variables del modelo (no exigimos year_surgery aquí)
dat_cc = dat.dropna(subset=[c for c in dat.columns if c != "year_surgery"])
dat_cc.to_csv(OUT_CC, index=False)
print(f"Guardado complete-case: {OUT_CC}  ({len(dat_cc)} filas × {len(dat_cc.columns)} columnas)")
print(f"Pacientes perdidos por missings: {len(dat) - len(dat_cc)}")
print(f"Prevalencia de cáncer en complete-case: {dat_cc['cancer'].mean()*100:.1f}%")

# ---------------------------------------------------------------------------
# 7. Cohorte para split temporal: requiere también year_surgery no-NA
# ---------------------------------------------------------------------------
OUT_TMP = ROOT / "analisis" / "datos_modelo_cc_temporal.csv"
dat_tmp = dat_cc.dropna(subset=["year_surgery"]).copy()
dat_tmp["cohort"] = (dat_tmp["year_surgery"] < 2019).map(
    {True: "train", False: "test"})
dat_tmp.to_csv(OUT_TMP, index=False)

print(f"\nGuardado split temporal: {OUT_TMP}  ({len(dat_tmp)} filas)")
print("\nCohorte temporal:")
print(dat_tmp.groupby(["cohort", "cancer"]).size().unstack(fill_value=0))
print("\nPrevalencia de cáncer por cohorte:")
print(dat_tmp.groupby("cohort")["cancer"].mean().mul(100).round(1))

File: src/01_preparar_dataset.R
# 01_preparar_dataset.R
# ---------------------------------------------------------------------
# Lee Calculadora 2024.sav, filtra CIRUGIA=1, selecciona las 13
# variables del modelo publicado + outcome (TODOSCA), recalcula age2 y
# guarda el dataset analítico como datos_modelo.rds (R) y .csv (Python).
#
# Equivalente exacto al script 01_preparar_dataset.py.
#
# Requiere: haven, dplyr
# ---------------------------------------------------------------------

suppressPackageStartupMessages({
  library(haven)
  library(dplyr)
})

ROOT <- here::here()  # o usar una ruta absoluta si no usas {here}
SAV  <- file.path(ROOT, "Calculadora 2024.sav")
OUT_RDS <- file.path(ROOT, "analisis", "datos_modelo.rds")
OUT_CSV <- file.path(ROOT, "analisis", "datos_modelo.csv")

# 1. Cargar -----------------------------------------------------------
df <- read_sav(SAV)
cat(sprintf("Cargado: %d pacientes x %d variables\n", nrow(df), ncol(df)))

# 2. Filtrar cohorte --------------------------------------------------
df <- df %>% filter(CIRUGIA == 1)
cat(sprintf("Tras filtrar CIRUGIA=1: %d pacientes\n", nrow(df)))

# 3. Variables del modelo + renombrado --------------------------------
dat <- df %>%
  transmute(
    fhx_tc           = AFCDT,
    male             = SEX,
    age              = EDAD,
    age2             = EDAD^2,                 # recalculado, SPSS tenía NAs
    tsh_low          = TSHSUPRIMIDA,
    tsh_high         = TSHALTA,
    autoimmune       = TIROIDITIS,
    solid            = SOLIDO1,
    susp_nodes       = GANGLIOSOSPECHOSO1,
    hypoechoic       = HIPOECOICO1,
    irregular        = IRREGULAR1,
    macrocalc        = MACROCA1,
    microcalc        = MICROCA1,
    taller_than_wide = FORMA1,
    cancer           = TODOSCA,
    year_surgery     = AÑOCIRUGIA
  ) %>%
  filter(!is.na(cancer)) %>%
  # Corregir typo evidente: año 2921 → 2021
  mutate(year_surgery = if_else(year_surgery == 2921, 2021, year_surgery))

# Convertir categóricas a factor (no a numéricas) para R/tidymodels
cat_vars <- c("fhx_tc","male","tsh_low","tsh_high","autoimmune",
              "solid","susp_nodes","hypoechoic","irregular",
              "macrocalc","microcalc","taller_than_wide","cancer")

dat <- dat %>%
  mutate(across(all_of(cat_vars), ~ factor(.x, levels = c(0, 1))))

# 4. Resumen ----------------------------------------------------------
cat("\nDistribución del outcome:\n")
print(table(dat$cancer, useNA = "always"))
cat(sprintf("Prevalencia de cáncer: %.1f%%\n",
            mean(dat$cancer == 1, na.rm = TRUE) * 100))

cat("\nMissings por variable:\n")
print(sort(colSums(is.na(dat)), decreasing = TRUE))

# 5. Guardar dos versiones: completa y complete-case ------------------
saveRDS(dat, OUT_RDS)
write.csv(dat, OUT_CSV, row.names = FALSE)

dat_cc <- na.omit(dat)
OUT_RDS_CC <- file.path(ROOT, "analisis", "datos_modelo_cc.rds")
OUT_CSV_CC <- file.path(ROOT, "analisis", "datos_modelo_cc.csv")
saveRDS(dat_cc, OUT_RDS_CC)
write.csv(dat_cc, OUT_CSV_CC, row.names = FALSE)

cat(sprintf("\nGuardado:\n  %s (n=%d)\n  %s (n=%d)\n",
            OUT_RDS, nrow(dat), OUT_RDS_CC, nrow(dat_cc)))
cat(sprintf("Pacientes perdidos por missings: %d\n", nrow(dat) - nrow(dat_cc)))
cat(sprintf("Prevalencia de cáncer en complete-case: %.1f%%\n",
            mean(dat_cc$cancer == 1) * 100))

# 6. Cohorte para split temporal: requiere también year_surgery no-NA -----
OUT_RDS_TMP <- file.path(ROOT, "analisis", "datos_modelo_cc_temporal.rds")
OUT_CSV_TMP <- file.path(ROOT, "analisis", "datos_modelo_cc_temporal.csv")

dat_tmp <- dat_cc %>%
  filter(!is.na(year_surgery)) %>%
  mutate(cohort = if_else(year_surgery < 2019, "train", "test"))

saveRDS(dat_tmp, OUT_RDS_TMP)
write.csv(dat_tmp, OUT_CSV_TMP, row.names = FALSE)

cat(sprintf("\nSplit temporal guardado: %s (n=%d)\n", OUT_RDS_TMP, nrow(dat_tmp)))
cat("\nTabla cohorte × cancer:\n")
print(table(dat_tmp$cohort, dat_tmp$cancer, useNA = "ifany"))
cat("\nPrevalencia de cáncer por cohorte (%):\n")
print(round(tapply(as.numeric(as.character(dat_tmp$cancer)),
                   dat_tmp$cohort, mean) * 100, 1))
